File: spi/src/test/java/com/github/ykiselev/spi/camera/FrustumTest.java
package com.github.ykiselev.spi.camera;

import com.github.ykiselev.opengl.matrices.Matrix;
import com.github.ykiselev.opengl.matrices.Vector3f;
import org.junit.jupiter.api.Test;

import java.nio.FloatBuffer;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class FrustumTest {

    private final Frustum frustum = new Frustum();

    @Test
    void shouldClassify() {
        FloatBuffer m = FloatBuffer.allocate(16);
        Matrix.perspective(-0.1f, 0.1f, 0.1f, -0.1f, 0.1f, 100, m);
        frustum.setFromMatrix(m);

        // check near plane
        assertTrue(frustum.isInside(new Vector3f(0, 0, -0.1f)));
        assertFalse(frustum.isInside(new Vector3f(0, 0, -0.05f)));
        assertTrue(frustum.intersects(new Vector3f(0, 0, -0.05f), 0.11f));

        // check far plane
        assertTrue(frustum.isInside(new Vector3f(0, 0, -100)));
        assertFalse(frustum.isInside(new Vector3f(0, 0, -100.5f)));

        // check left plane
        assertTrue(frustum.isInside(new Vector3f(-0.1f, 0, -0.1f)));
        assertFalse(frustum.isInside(new Vector3f(-0.11f, 0, -0.1f)));

        // check right plane
        assertTrue(frustum.isInside(new Vector3f(0.1f, 0, -0.1f)));
        assertFalse(frustum.isInside(new Vector3f(0.11f, 0, -0.1f)));

        // check top plane
        assertTrue(frustum.isInside(new Vector3f(0, 0.1f, -0.1f)));
        assertFalse(frustum.isInside(new Vector3f(0, 0.11f, -0.1f)));

        // check bottom plane
        assertTrue(frustum.isInside(new Vector3f(0, -0.1f, -0.1f)));
        assertFalse(frustum.isInside(new Vector3f(0, -0.11f, -0.1f)));

        // check point inside
        assertTrue(frustum.isInside(new Vector3f(0, 0, -1)));
    }
}

File: spi/src/test/java/com/github/ykiselev/spi/camera/FrustumTest.kt
package com.github.ykiselev.spi.camera

import com.github.ykiselev.opengl.matrices.Vector3f
import com.github.ykiselev.opengl.matrices.perspective
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.nio.FloatBuffer

internal class FrustumTest {

    private val frustum = Frustum()

    @Test
    fun shouldClassify() {
        val m = FloatBuffer.allocate(16)

        perspective(-0.1f, 0.1f, 0.1f, -0.1f, 0.1f, 100f, m)

        frustum.setFromMatrix(m)

        // check near plane
        Assertions.assertTrue(frustum.isInside(Vector3f(0f, 0f, -0.1f)))
        Assertions.assertFalse(frustum.isInside(Vector3f(0f, 0f, -0.05f)))
        Assertions.assertTrue(frustum.intersects(Vector3f(0f, 0f, -0.05f), 0.11f))

        // check far plane
        Assertions.assertTrue(frustum.isInside(Vector3f(0f, 0f, -100f)))
        Assertions.assertFalse(frustum.isInside(Vector3f(0f, 0f, -100.5f)))

        // check left plane
        Assertions.assertTrue(frustum.isInside(Vector3f(-0.1f, 0f, -0.1f)))
        Assertions.assertFalse(frustum.isInside(Vector3f(-0.11f, 0f, -0.1f)))

        // check right plane
        Assertions.assertTrue(frustum.isInside(Vector3f(0.1f, 0f, -0.1f)))
        Assertions.assertFalse(frustum.isInside(Vector3f(0.11f, 0f, -0.1f)))

        // check top plane
        Assertions.assertTrue(frustum.isInside(Vector3f(0f, 0.1f, -0.1f)))
        Assertions.assertFalse(frustum.isInside(Vector3f(0f, 0.11f, -0.1f)))

        // check bottom plane
        Assertions.assertTrue(frustum.isInside(Vector3f(0f, -0.1f, -0.1f)))
        Assertions.assertFalse(frustum.isInside(Vector3f(0f, -0.11f, -0.1f)))

        // check point inside
        Assertions.assertTrue(frustum.isInside(Vector3f(0f, 0f, -1f)))
    }
}